File: bindings/python/perf/single/perf_pubsub.py
import sys
import threading
import time

import zlink

from perf_common import (
    STOP_TOKEN,
    apply_single_auto_hwm_msg_unit,
    apply_single_socket_options,
    benchmark_run_id,
    configure_single_tls_client,
    configure_single_tls_server,
    new_payload,
    parse_single_args,
    perf_context,
    poll_idle_ms,
    print_result_lines,
    resolve_single_endpoint,
    resolve_single_connect_ready_timeout_ms,
    resolve_single_pubsub_ready_settle_s,
    resolve_single_pubsub_recv_timeout_ms,
    result_metrics,
    run_one_way_subscriber_public_subscribe,
    stamp_payload,
    wait_monitor_event,
)


TOPIC = b"bench"


def _publish_stop_token(publisher):
    """PERF_SINGLE_TEST_POLICY § 1.4 wire-level shutdown signal."""

    for _ in range(100):
        try:
            publisher.publish(TOPIC).message(STOP_TOKEN).submit()
            return
        except zlink.SubmitError as exc:
            if exc.result != zlink.SubmitResult.BACKPRESSURED:
                raise
            poll_idle_ms(1)


def _public_one_way_metrics(sender, receiver, *, msg_size, duration_s, run_id):
    return None


def main(argv=None):
    args = parse_single_args(argv or sys.argv[1:], pattern="pubsub")
    payload = new_payload(args.msg_size)
    run_id = benchmark_run_id()
    latencies = []
    received = 0

    def send_loop(publisher, active_end):
        # C send_active_samples: DONTWAIT publish, re-stamp fresh now_ns on
        # every retry, busy-loop through transient backpressure.
        flag = int(zlink.SendFlags.DONT_WAIT)
        publish = publisher.publish
        topic = TOPIC
        stamp = stamp_payload
        perf_counter = time.perf_counter
        submit_backpressured = zlink.SubmitResult.BACKPRESSURED
        while perf_counter() < active_end:
            try:
                publish(topic).message(stamp(payload, phase=1, run_id=run_id)).flags(
                    flag
                ).submit()
            except zlink.SubmitError as exc:
                if exc.result != submit_backpressured:
                    raise
        _publish_stop_token(publisher)
    with perf_context() as ctx:
        apply_single_auto_hwm_msg_unit(ctx, args.msg_size)
        with zlink.create_pub_socket(ctx) as publisher:
            # Match C perf: the harness always sets NODROP explicitly. The
            # socket default is lossy fanout, which would drop samples and the
            # stop token.
            publisher.pub_options.no_drop = True
            with zlink.create_sub_socket(ctx) as subscriber:
                endpoint = resolve_single_endpoint(args.transport, "pubsub")
                apply_single_socket_options(
                    publisher,
                    subscriber,
                    receive_timeout_ms=resolve_single_pubsub_recv_timeout_ms(),
                )
                configure_single_tls_server(publisher, args.transport)
                configure_single_tls_client(subscriber, args.transport)
                subscriber.set_subscription(TOPIC)
                with subscriber.monitor_open(zlink.MonitorEventMask.CONNECTION_READY) as monitor:
                    publisher.bind(endpoint)
                    subscriber.connect(endpoint)
                    wait_monitor_event(
                        monitor,
                        zlink.MonitorEventMask.CONNECTION_READY,
                        timeout_ms=resolve_single_connect_ready_timeout_ms(),
                    )
                wait_seconds = resolve_single_pubsub_ready_settle_s()
                if wait_seconds > 0:
                    time.sleep(wait_seconds)

                active_end = time.perf_counter() + args.duration
                metrics = _public_one_way_metrics(
                    publisher,
                    subscriber,
                    msg_size=args.msg_size,
                    duration_s=args.duration,
                    run_id=run_id,
                )
                if metrics is None:
                    sender = threading.Thread(
                        target=send_loop, args=(publisher, active_end), daemon=True
                    )
                    sender.start()
                    # C perf_pubsub.cpp run_active_phase receiver.
                    received = run_one_way_subscriber_public_subscribe(
                        subscriber,
                        topic=TOPIC,
                        msg_size=args.msg_size,
                        run_id=run_id,
                        active_end=active_end,
                        received=received,
                        latencies=latencies,
                    )

                    sender.join()
                    if received == 0:
                        raise RuntimeError(
                            "pubsub benchmark did not receive any active message"
                        )
                    metrics = result_metrics(
                        count=received,
                        msg_size=args.msg_size,
                        elapsed_s=args.duration,
                        latencies_ns=latencies,
                    )
                print_result_lines("PUBSUB", args.transport, args.msg_size, metrics)


if __name__ == "__main__":
    main()

File: bindings/rust/perf/single/src/perf_pubsub.rs
//! Single PUB/SUB throughput/latency benchmark.

mod common;
mod one_way;

use zlink::{Message, RecvFlags, SocketMonitor, SubmitResult, TopicMessage};

fn main() {
    let config = common::PerfConfig::from_env_and_args();
    let Some(bind_endpoint) =
        common::resolve_endpoint_or_emit_unsupported("PUBSUB", &config.transport, "pubsub")
    else {
        return;
    };

    let ctx = common::perf_context();
    let pub_sock = ctx.pub_socket().expect("pub");
    let sub_sock = ctx.sub_socket().expect("sub");
    // Match C perf: the harness always sets NODROP explicitly. The socket
    // default is lossy fanout, which would drop samples and the stop token.
    pub_sock
        .pub_options()
        .set_no_drop(true)
        .expect("pub no_drop");
    // Match C perf: numeric socket HWM remains behind the manual-override gate.
    common::apply_single_hwm(&pub_sock);
    common::apply_single_hwm(&sub_sock);
    if matches!(config.transport.as_str(), "tls" | "wss") {
        let tls = common::resolve_perf_tls_paths().expect("TLS certs not found");
        common::setup_raw_tls_server(&pub_sock, &tls).expect("pub tls");
        common::setup_raw_tls_client(&sub_sock, &tls).expect("sub tls");
    }

    let mut pub_mon = SocketMonitor::open(&pub_sock).expect("pub monitor");
    let mut mon = SocketMonitor::open(&sub_sock).expect("monitor");
    if let Err(err) = pub_sock.bind(&bind_endpoint) {
        if common::handle_transport_setup_error("PUBSUB", &config.transport, "bind", err) {
            return;
        }
        panic!("bind: {err}");
    }
    let endpoint = pub_sock.last_endpoint().unwrap_or(bind_endpoint);
    if let Err(err) = sub_sock.connect(&endpoint) {
        if common::handle_transport_setup_error("PUBSUB", &config.transport, "connect", err) {
            return;
        }
        panic!("connect: {err}");
    }
    sub_sock.set_subscription("").expect("subscribe");
    let ready_timeout = common::resolve_single_ready_timeout();
    common::wait_monitor_ready(&mut pub_mon, ready_timeout, "pubsub publisher");
    common::wait_monitor_ready(&mut mon, ready_timeout, "pubsub subscriber");
    std::thread::sleep(common::resolve_single_pubsub_ready_settle());

    let collector = common::MetricCollector::new();
    let stats = collector.shared();

    let active = std::time::Duration::from_secs(config.duration_seconds);
    let active_deadline = common::now_ns() + active.as_nanos() as u64;
    let send_thread = std::thread::spawn(move || {
        common::send_loop(active_deadline, config.size, common::PHASE_ACTIVE, |msg| {
            match if common::measurement_part_count() == 2 {
                pub_sock
                    .publish("P")
                    .message(msg)
                    .message(Message::try_from(&[] as &[u8]).expect("empty measurement tail"))
                    .flags(zlink::SendFlags::DONT_WAIT)
                    .submit()
            } else {
                pub_sock
                    .publish("P")
                    .message(msg)
                    .flags(zlink::SendFlags::DONT_WAIT)
                    .submit()
            } {
                Ok(()) => true,
                Err(err)
                    if err.code() == SubmitResult::NotConnected
                        || common::is_single_send_retry_error(&err) =>
                {
                    false
                }
                Err(err) => panic!("active publish: {err}"),
            }
        });
        common::send_stop_token(|msg| pub_sock.publish("P").message(msg).submit().map(|()| true));
        pub_sock
    });

    let mut received = TopicMessage::empty();
    one_way::recv_until_stop(&sub_sock, || {
        match sub_sock.subscribe(&mut received, RecvFlags::DONT_WAIT) {
            Ok(true) => {
                let data = match one_way::classify(received.parts()) {
                    one_way::Content::Payload(data) => data,
                    one_way::Content::Stop => return one_way::RecvStep::Stop,
                };
                common::handle_recv(data, config.size, &stats, active_deadline);
                one_way::RecvStep::Payload
            }
            Ok(false) => one_way::RecvStep::Empty,
            Err(err) => panic!("pubsub subscriber recv failed: {err}"),
        }
    });
    let _pub_sock = send_thread.join().expect("sender thread");

    let result = collector.finish();
    common::print_result(
        "PUBSUB",
        &config.transport,
        config.size,
        config.duration_seconds,
        &result,
    );

    // Match the C guard destruction order: disconnect the subscriber before
    // closing the inproc publisher endpoint.
    drop(mon);
    drop(pub_mon);
    drop(sub_sock);
    drop(_pub_sock);
}
